### ISSUE TITLE
state resets still happen in v2 rooms

### ISSUE THREAD

**Issue by richvdh** [matrix-org/synapse#8629]
I just upgraded `#synapse:matrix.org` because I got fed up of losing my ops bit. It was a v4 room.

**Comment by BillCarsonFr** [comment]
Just to clarify, that means that it still happens in v2+ (3...9) versions right?

**Comment by Mikaela** [comment]
Yes, since the state resolution algorithm hasn't been changed since room version 2

* https://spec.matrix.org/v1.1/rooms/#complete-list-of-room-versions

**Comment by jae1911** [comment]
Hey there,
Every now and then in Furry Tech (`!vRGLvqJYlFvzpThbxI:matrix.org`; room v6) we are getting a state reset that mostly resets profile pictures and usernames.

**Comment by ptman** [comment]
I was under the impression that state res v2 was studied and deemed sound. https://matrix.org/blog/2020/06/16/matrix-decomposition-an-independent-academic-analysis-of-matrix-state-resolution . Did the research group finish work on a formal proof?

**Comment by ara4n** [comment]
There's a large project within the core team currently working through a new formal expression of state res v2 to try to characterise the remaining issues.  Given the risk of discovering security issues, it's being treated as an internal security project (juggled alongside our other security work). There's been some progress in identifying the root cause here though; will update as soon we can.

**Comment by DMRobertson** [comment]
> Hey there,
> Every now and then in Furry Tech (`!vRGLvqJYlFvzpThbxI:matrix.org`; room v6) we are getting a state reset that mostly resets profile pictures and usernames.

Hi @jae1911. I would like to look into the details here. Could you share

- an example of a user whose profile was reset
- a time range before and after the reset (with timezone, if relevant)
- (if possible) the event id of that user's `m.room.join` event before and after the reset?

**Comment by jae1911** [comment]
> > Hey there,
> > Every now and then in Furry Tech (`!vRGLvqJYlFvzpThbxI:matrix.org`; room v6) we are getting a state reset that mostly resets profile pictures and usernames.
> 
> Hi @jae1911. I would like to look into the details here. Could you share
> 
>     * an example of a user whose profile was reset
> 
>     * a time range before and after the reset (with timezone, if relevant)
> 
>     * (if possible) the event id of that user's `m.room.member` event before and after the reset?

Hey there, sure!  
I will update this comment when a state reset reoccurs.  
For now, I will take my own user `@me:jae.fi`, that is getting reset every now and then, taken 2022/07/28 21:56 (Helsinki Time):
```
{
  "type": "m.room.member",
  "sender": "@me:jae.fi",
  "content": {
    "membership": "join",
    "displayname": "Jae the bread lover",
    "avatar_url": "mxc://jae.fi/c5e04109ec88975b4a4b9b3fc73c06716a5a235e"
  },
  "state_key": "@me:jae.fi",
  "origin_server_ts": 1658996393543,
  "unsigned": {
    "replaces_state": "$qK0u0h-fP2Kd_koNW2nkddH8VC-rJ7M8KGiCXmN4lIY",
    "prev_content": {
      "membership": "join",
      "displayname": "Jae the bread lover",
      "avatar_url": "mxc://jae.fi/d31d8316f1342c4e44aa599262d0500dbcde0f67"
    },
    "prev_sender": "@me:jae.fi",
    "age": 28925483
  },
  "event_id": "$B9d_2RBfpymx6eyq7MPArpwZcHDbv7yFc7s-Nj97LbE",
  "room_id": "!vRGLvqJYlFvzpThbxI:matrix.org"
}
```

**Comment by DMRobertson** [comment]
> Ping @DMRobertson ; state reset occurred in the night of the 2022/08/03; here is the `m.room.member` of my user at 06:12;

@jae1911 I assume this is from the perspective of `jae.fi`? From the `matrix.org` perspective, your membership is currently `$RD1ZT-xYbd8WqjqqktsvKyyRlI8dnzWf3kVL7ojOAwA` (created by `jae.fi` at `2022-08-02 15:11:21+01`).

Are you able to run sql queries on the Synapse database hosting `jae.fi`? If so, please reach out to `@dmrobertson:matrix.org`---I'd like to ask you to run a few SELECTs to establish what prompted the reset.

**Comment by DMRobertson** [comment]
Thanks for getting in touch @jae1911. From the queries we ran, we saw that from the point of view of `jae.fi`:

- your membership of the room recently changed from `$GXG0iXSO6ICGZ_XoLr_rSR_nuDmrKjBGfbAliDhrDic` to `$SH9zmA3rnQutHvL2GGxfJmEsnaatPB-7Y-cTzqL80Pc` to `$RD1ZT-xYbd8WqjqqktsvKyyRlI8dnzWf3kVL7ojOAwA`
- the current state of the room has your membership determined by `$RD1ZT-xYbd8WqjqqktsvKyyRlI8dnzWf3kVL7ojOAwA`
- a new account also saw that your membership was determined by `$RD1ZT-xYbd8WqjqqktsvKyyRlI8dnzWf3kVL7ojOAwA`

The first two bullets are identical to what I see on `matrix.org`.

Therefore, I don't think this is an example of a server-side state reset; rather, I think this is an instance of #4508 (or something like it) where clients can have their view of state differ from that of the server. Debugging that problem will be fiddly and should hopefully be avoided by using sliding sync rather than that current sync API.

With that said, please do let me know if you see another instance of state that appears to have reset unexpectedly.

I will edit the description of this to include the relevant sql queries for any future reports.

**Comment by Mikaela** [comment]
This happened to be in the Furry Tech room. Looking from Nheko on pikaviestin.fi my avatar and name went back in time to `Pikaminda ⚡` from `Dea Mortis`. `jae.fi` sees the expected profile though.

**Comment by DMRobertson** [comment]
> This happened to be in the Furry Tech room. Looking from Nheko on pikaviestin.fi my avatar and name went back in time to Pikaminda ⚡ from Dea Mortis. jae.fi sees the expected profile though.

I don't see the reset in matrix.org's database. Are you able to run the SQL queries in the description?

**Comment by Mikaela** [comment]
I am not a server admin anywhere

**Comment by JokerGermany** [comment]
This Message was deleted but can still be seen on matrix.org
https://matrix.to/#/!POnsZmdSAzRrxMaTOj:matrix.kraut.space/$JokvE8q-ggX02mw-r4Lv3PvW2Tkh431WBrXxfJ5CFWo?via=behm.me&via=matrix.org&via=tchncs.de

**Comment by DMRobertson** [comment]
> This Message was deleted but can still be seen on matrix.org
> They are in sync now, the message is deleted.

Sounds like there was some delay in the redaction event `$O4n91n4jwoQZH7nB0w-7ySZ_PFQVx5OlM4ljcH56wnY` sent by `alea.gnuu.de` reading `matrix.org`. Given that this resolved itself I don't think this is related to a state reset.

**Comment by vivia** [comment]
I had created `#randomstuffhere:gnome.org` and set up a few co-admins together with myself. Now I just realised we've all been demoted to moderators, and I don't have the admin rights to fix that myself. I asked for help on `#element-web:matrix.org` and was told to write here.

**Comment by vivia** [comment]
Thanks, this actually adds up with what I remember happening. We had a different room in the past, and at some point several months ago, the IRC bridge was acting up and required bridge admin intervention. When trying to fix this, I moved us all to this room that had been created by that specific creator (staying intentionally vague here, but I'm sure you will understand exactly what I mean). Apparently in the process our power levels became Moderators instead, and I never noticed until now because that happens to give me enough privileges to manage a low-traffic room with just friends. I guess it can stay like this then, people can get moderator status by opping themselves on the IRC side, where I can manage ChanServ.

Thanks for your investigation!

**Comment by Mikaela** [comment]
We at `#fluffychat:matrix.org` are getting reports of the room being DAG split and thus view of avatars (including the room avatar) depending on which homeserver is being used.

Can this situation be confirmed somehow and that room upgrade is absolutely necessary or is knowledge of it at all helpful? I understand that DAG splits are also tracked in this issue.

**Comment by DMRobertson** [comment]
> Can this situation be confirmed somehow

We'd need more details (see also the top comment, which I edited some months back with some suggestions). Whose avatar is being split? Which servers see it differently; ideally, which event ID do they report as being current?

**Comment by Mikaela** [comment]
* Room ID is `!IwPxifXSjBLghpVEMh:matrix.org`
* Room avatar is missing from some servers
* Users such as `@krille:janian.de` and `@mxidupwitch:the-apothecary.club` have avatars missing depending on server. I cannot see the later or their display name on `pikaviestin.fi` using Nheko
* at least matrix.org and pikaviestin.fi and the-apothecary.club and arcticfoxes.net are reported seeing different state and the last has been claiming multiple DAG splits past few months and upgraded most if not all of their rooms.

I don't know how to see current event ID. I have no access to any database personally. 

I am going by assumption that it's better that issues like this are reported rather than letting them happen silently and migrating to protocols where they don't happen.

DAG split was claimed in event https://matrix.to/#/%23fluffychat%3Amatrix.org/%24DN3Au8zP9FMwDXULg_ZJZmXkw5BcKVDCL-BHx3-ARvM?via=pikaviestin.fi&via=matrix.org&via=the-apothecary.club&via=envs.net

**Comment by wj25czxj47bu6q** [comment]
I have documented a new case of state reset at https://github.com/matrix-org/synapse/issues/15987#issuecomment-1655485693.

**Comment by wj25czxj47bu6q** [comment]
I have documented a new case of state reset at https://github.com/matrix-org/synapse/issues/16629.
